File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Symbols/SingleNamespaceDeclaration.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using Microsoft.CodeAnalysis.VisualBasic.Syntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
{
	internal class SingleNamespaceDeclaration : SingleNamespaceOrTypeDeclaration
	{
		private class Comparer : IEqualityComparer<SingleNamespaceDeclaration>
		{
			private bool Equals(SingleNamespaceDeclaration decl1, SingleNamespaceDeclaration decl2)
			{
				return CaseInsensitiveComparison.Equals(decl1.Name, decl2.Name);
			}

			bool IEqualityComparer<SingleNamespaceDeclaration>.Equals(SingleNamespaceDeclaration decl1, SingleNamespaceDeclaration decl2)
			{
				//ILSpy generated this explicit interface implementation from .override directive in Equals
				return this.Equals(decl1, decl2);
			}

			private int GetHashCode(SingleNamespaceDeclaration decl1)
			{
				return CaseInsensitiveComparison.GetHashCode(decl1.Name);
			}

			int IEqualityComparer<SingleNamespaceDeclaration>.GetHashCode(SingleNamespaceDeclaration decl1)
			{
				//ILSpy generated this explicit interface implementation from .override directive in GetHashCode
				return this.GetHashCode(decl1);
			}
		}

		private readonly ImmutableArray<SingleNamespaceOrTypeDeclaration> _children;

		public readonly bool IsPartOfRootNamespace;

		public static readonly IEqualityComparer<SingleNamespaceDeclaration> EqualityComparer = new Comparer();

		public bool HasImports { get; set; }

		public virtual bool IsGlobalNamespace => false;

		public override DeclarationKind Kind => DeclarationKind.Namespace;

		public SingleNamespaceDeclaration(string name, bool hasImports, SyntaxReference syntaxReference, Location nameLocation, ImmutableArray<SingleNamespaceOrTypeDeclaration> children, bool isPartOfRootNamespace = false)
			: base(name, syntaxReference, nameLocation)
		{
			_children = children;
			HasImports = hasImports;
			IsPartOfRootNamespace = isPartOfRootNamespace;
		}

		protected override ImmutableArray<SingleNamespaceOrTypeDeclaration> GetNamespaceOrTypeDeclarationChildren()
		{
			return _children;
		}

		public NamespaceBlockSyntax GetNamespaceBlockSyntax()
		{
			if (SyntaxReference == null)
			{
				return null;
			}
			return SyntaxReference.GetSyntax().AncestorsAndSelf().OfType<NamespaceBlockSyntax>()
				.FirstOrDefault();
		}

		public new static string BestName<T>(ImmutableArray<T> singleDeclarations, ref bool multipleSpellings) where T : SingleNamespaceDeclaration
		{
			multipleSpellings = false;
			string text = singleDeclarations[0].Name;
			int num = singleDeclarations.Length - 1;
			for (int i = 1; i <= num; i++)
			{
				string name = singleDeclarations[i].Name;
				int num2 = string.Compare(text, name, StringComparison.Ordinal);
				if (num2 != 0)
				{
					multipleSpellings = true;
					if (singleDeclarations[0].IsPartOfRootNamespace)
					{
						return text;
					}
					if (singleDeclarations[i].IsPartOfRootNamespace)
					{
						return name;
					}
					if (num2 > 0)
					{
						text = name;
					}
				}
			}
			return text;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Symbols/SingleNamespaceDeclaration.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports System
Imports System.Collections.Generic
Imports System.Collections.Immutable
Imports System.Linq
Imports System.Runtime.CompilerServices
Imports System.Threading

Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
	Friend Class SingleNamespaceDeclaration
		Inherits SingleNamespaceOrTypeDeclaration
		Private ReadOnly _children As ImmutableArray(Of SingleNamespaceOrTypeDeclaration)

		Public ReadOnly IsPartOfRootNamespace As Boolean

		Public ReadOnly Shared EqualityComparer As IEqualityComparer(Of SingleNamespaceDeclaration)

		Public Property HasImports As Boolean

		Public Overridable ReadOnly Property IsGlobalNamespace As Boolean
			Get
				Return False
			End Get
		End Property

		Public Overrides ReadOnly Property Kind As DeclarationKind
			Get
				Return DeclarationKind.[Namespace]
			End Get
		End Property

		Shared Sub New()
			SingleNamespaceDeclaration.EqualityComparer = New SingleNamespaceDeclaration.Comparer()
		End Sub

		Public Sub New(ByVal name As String, ByVal hasImports As Boolean, ByVal syntaxReference As Microsoft.CodeAnalysis.SyntaxReference, ByVal nameLocation As Microsoft.CodeAnalysis.Location, ByVal children As ImmutableArray(Of SingleNamespaceOrTypeDeclaration), Optional ByVal isPartOfRootNamespace As Boolean = False)
			MyBase.New(name, syntaxReference, nameLocation)
			Me._children = children
			Me.HasImports = hasImports
			Me.IsPartOfRootNamespace = isPartOfRootNamespace
		End Sub

		Public Shared Shadows Function BestName(Of T As SingleNamespaceDeclaration)(ByVal singleDeclarations As ImmutableArray(Of T), ByRef multipleSpellings As Boolean) As String
			Dim str As String
			multipleSpellings = False
			Dim name As String = singleDeclarations(0).Name
			Dim length As Integer = singleDeclarations.Length - 1
			Dim num As Integer = 1
			While True
				If (num <= length) Then
					Dim name1 As String = singleDeclarations(num).Name
					Dim num1 As Integer = [String].Compare(name, name1, StringComparison.Ordinal)
					If (num1 <> 0) Then
						multipleSpellings = True
						If (singleDeclarations(0).IsPartOfRootNamespace) Then
							str = name
							Exit While
						ElseIf (singleDeclarations(num).IsPartOfRootNamespace) Then
							str = name1
							Exit While
						ElseIf (num1 > 0) Then
							name = name1
						End If
					End If
					num = num + 1
				Else
					str = name
					Exit While
				End If
			End While
			Return str
		End Function

		Public Function GetNamespaceBlockSyntax() As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax
			Dim namespaceBlockSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax
			If (Me.SyntaxReference IsNot Nothing) Then
				Dim syntaxReference As Microsoft.CodeAnalysis.SyntaxReference = Me.SyntaxReference
				Dim cancellationToken As System.Threading.CancellationToken = New System.Threading.CancellationToken()
				namespaceBlockSyntax = syntaxReference.GetSyntax(cancellationToken).AncestorsAndSelf(True).OfType(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax)().FirstOrDefault()
			Else
				namespaceBlockSyntax = Nothing
			End If
			Return namespaceBlockSyntax
		End Function

		Protected Overrides Function GetNamespaceOrTypeDeclarationChildren() As ImmutableArray(Of SingleNamespaceOrTypeDeclaration)
			Return Me._children
		End Function

		Private Class Comparer
			Implements IEqualityComparer(Of SingleNamespaceDeclaration)
			Public Sub New()
				MyBase.New()
			End Sub

			Private Function Equals(ByVal decl1 As SingleNamespaceDeclaration, ByVal decl2 As SingleNamespaceDeclaration) As Boolean Implements IEqualityComparer(Of SingleNamespaceDeclaration).Equals
				Return CaseInsensitiveComparison.Equals(decl1.Name, decl2.Name)
			End Function

			Private Function GetHashCode(ByVal decl1 As SingleNamespaceDeclaration) As Integer Implements IEqualityComparer(Of SingleNamespaceDeclaration).GetHashCode
				Return CaseInsensitiveComparison.GetHashCode(decl1.Name)
			End Function
		End Class
	End Class
End Namespace